docs/desired-state-diff/index.md

```diff
@@ -0,0 +1,203 @@
+      "type": "Modified", // "Added" | "Deleted" | "Modified"
+      "attributes": [
+        {
+          "id": "purged", // string
+          "type": "Modified", // "Modified" | "Unmodified"
+          "diff": "--- \n+++ \n@@ -0,0 +1 @@\n+false", // unified diff
+          "source": false, // old value
+          "target": false // new value
+        },
+        {
+          "id": "peer_ip", // string
+          "type": "Unmodified", // "Modified" | "Unmodified"
+          "diff": null, // unified diff
+          "source": "10.255.136.9", // old value
+          "target": "10.255.136.9" // new value
+        }
+      ]
+    }
+  ]
+}
+```
+
+If we don't show unchanged attributes in the frontend, there is no point in having the type property on attributes ("Modified" | "Unmodified") as all of them are Modified.
+
+The unified diff string should contain filenames.  
+I am not sure if this is required for the official format.
+But the `diff2html` library does not work without it.
+Other libraries don't seem to care.
+
+Does not work:
+
+```
+---
++++
+@@ -0,0 +1 @@
++data-network-internet1-1
+```
+
+Works:
+
+```
+--- a
++++ a
+@@ -0,0 +1 @@
++data-network-internet1-1
+```
+
+If we want a resource overview, and the data has pagination, then we need to get this list of resource ids somehow.  
+Maybe a different endpoint?
+
+## Diff Highlight
+
+As far as I am aware, the unified diff format does not contain inline changes, only line changes.  
+This means to highlight the actual change on a single line, the line has to be diffed in the frontend.
+
+Example:
+
+```
+---
++++
+@@ -1 +1 @@
+-data-network-internet1-1
++data-network-internet1-2
+```
+
+There is only 1 character change, but the line is labeled as "Deleted" and "Added".  
+Some frontend libraries do a diff and highlight the actual changed character.
+
+## Conclusion
+
+There are some libraries that seem useful. Some are quite outdated and unmaintained. Best to avoid those.  
+Having the unified diff done on the backend seems the right approach. But some libraries do the diffing internally using another library.
+And single line highlight is always done on the frontend.
+
+The backend is responsible for listing the resources and their changes.  
+The actual diff of 2 attributes can be done in the frontend.  
+But the backend should provide a string representation of the attribute values.  
+The layout and containers should always be built by us.  
+For the actuall diff visual, we have 2 options:
+
+### Option 1, build the visual ourselves
+
+We have complete freedom to do what we want. But we might not need it.  
+We can use `jsdiff` to generate a json structure describing the unchanged, deleted and added characters.
+
+### Option 2, use a lib for generating the visual
+
+We can use a component like `react-diff-viewer`. There are some options. But we will always be limited.  
```

Comment:
For me, the question is that is there a library that's good enough for our purposes?

Reply:
I think react-diff-viewer might be good enough. It's not really well maintained.
But as far as I can see, its got everything covered.